python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_estimator.py

```diff
@@ -158,15 +158,25 @@ def __init__(
 
         local_init_params = self.worker_init_params.copy()
         local_init_params["log_to_driver"] = False
+        self.tcp_port = find_free_port()
+        self.num_executors = int(sc.getConf().get("spark.executor.instances"))
         self.driver_runner = PytorchPysparkWorker(
             mode='predict',
             cluster_info=self._get_cluster_info(sc),
             **local_init_params)
 
         self.state_dict = self.driver_runner.get_state_dict()
 
-    def _get_cluster_info(self, sc):
-        cluster_info = self.workerRDD.barrier().mapPartitions(find_ip_and_free_port).collect()
+    def create_tcpstore_server(self):
+        import torch.distributed as dist
+        from torch._C._distributed_c10d import _DEFAULT_PG_TIMEOUT
+        server_store = dist.TCPStore(self.ip, self.tcp_port, -1, True, _DEFAULT_PG_TIMEOUT)
+        return server_store
+
+    def _get_cluster_info(self):
+        cluster_info = []
+        for i in range(self.num_executors):
+            cluster_info.append(f"{self.ip}:{self.tcp_port}")
```

Comment:
It seems all workers get the same `ip:port` in `cluster_info`. Since we infer the world rank for each worker through `cluster_info`, that may lead to different workers sharing the same rank. I wonder whether PyTorch ddp can work correctly in that case...

Reply:
But workers should get the same `ip:port`, I will add a new function to deal with rank.

Reply:
Why are we using the driver `ip:port` in the `cluster_info`? I would assume `cluster_info` contains the information of all the PyTorch workers?

Reply:
> Why are we using the driver `ip:port` in the `cluster_info`? I would assume `cluster_info` contains the information of all the PyTorch workers?

TCPStore is a server-client, when the server store is created on driver, we need to set other client stores' URL and Port the same as the server store. Or it will cause a connection timeout error.

Reply:
> Why are we using the driver `ip:port` in the `cluster_info`? I would assume `cluster_info` contains the information of all the PyTorch workers?

We cannot use each worker's `ip:port` to create tcpstore, that's not a work solution.

Reply:
@sgwhat Shall we keep the original `cluster_info`, and maybe directly pass the `driver_ip` and `tcp_port` to workers?

Reply:
> @sgwhat Shall we keep the original `cluster_info`, and maybe directly pass the `driver_ip` and `tcp_port` to workers?

Replacing the `cluster_info` is one of the PR's purpose, since the original `cluster_info` is not working on k8s.

Reply:
> 

We could no longer use `cluster_info`, the key objective is to make sure each executor could reach the `ip:port` that we used to create server store on the driver.

Reply:
> > > @sgwhat Shall we keep the original `cluster_info`, and maybe directly pass the `driver_ip` and `tcp_port` to workers?
> > 
> > 
> > Replacing the `cluster_info` is one of the PR's purpose, since the original `cluster_info` is not working on k8s.
> 
> How is `cluster_info` used after the changes? It does not make sense to have a `cluster_info` with all the same `ip:port`. Maybe you no longer need a `cluster_info` after the changes.

We could no longer use `cluster_info`, the key objective is to make sure each executor could reach the `ip:port` that we used to create server store on the driver.

Reply:
`cluster_info` is also used to infer **rank** for each worker. You need to fix it as we mentioned before, the identical `ip:port` results in the same rank for different workers. Also, testing with more than 2 pods should uncover the issue. 

Replacing `cluster_info` might be a way but not the purpose of this PR. Before the change, we used `cluster_info[0]` as an identifier URL for pytorch DDP (`dist.init_processs_group(init_method=cluster_info[0])`). You could just change the URL to `{driver_ip}:{port}` to achieve your purpose, instead of directly changing the semantics of `cluster_info`.





https://github.com/intel-analytics/BigDL/blob/0d6d445eed986379222a55c7f0e84b0c29631a99/python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_worker.py#L114-L118

Reply:
> `cluster_info` is also used to infer **rank** for each worker. You need to fix it as we mentioned before, the identical `ip:port` results in the same rank for different workers. Also, testing with more than 2 pods should uncover the issue.
> 
> Replacing `cluster_info` might be a way but not the purpose of this PR. Before the change, we used `cluster_info[0]` as an identifier URL for pytorch DDP (`dist.init_processs_group(init_method=cluster_info[0])`). You could just change the URL to `{driver_ip}:{port}` to achieve your purpose, instead of directly changing the semantics of `cluster_info`.
> 
> https://github.com/intel-analytics/BigDL/blob/0d6d445eed986379222a55c7f0e84b0c29631a99/python/orca/src/bigdl/orca/learn/pytorch/pytorch_pyspark_worker.py#L114-L118

It's a better solution than replacing `cluster_info` directly.

Reply:
> It's a better solution than replacing cluster_info directly.

Great!